lcm-derive/src/parse.rs

```diff
@@ -0,0 +1,322 @@
+{
+	let segs = match *t {
+		syn::Type::Path(syn::TypePath { path: syn::Path { ref segments, .. }, ..}) => segments,
+		_ => panic!("Bug: `get_vec_inner_type` called on non-Vec type (1)"),
+	};
+
+	match *segs.iter().last().unwrap() {
+		syn::PathSegment {
+			arguments: syn::PathArguments::AngleBracketed(
+				syn::AngleBracketedGenericArguments { ref args, .. }
+			),
+			..
+		} => {
+			if let syn::GenericArgument::Type(ref ty) = args[0] {
+				ty
+			} else { panic!("Bug: `get_vec_inner_type` called on non-Vec type (2)"); }
+		},
+		_ => panic!("Bug: `get_vec_inner_type` called on non-Vec type (3)")
+	}
+}
+
+#[derive(Clone, Debug)]
+pub enum Ty
+{
+	Int8,
+	Int16,
+	Int32,
+	Int64,
+	Float,
+	Double,
+	String,
+	Boolean,
+	User(String)
+}
+impl Ty
+{
+	pub fn is_primitive_type(&self) -> bool
+	{
+		match *self {
+			Ty::User(_) => false,
+			_           => true
+		}
+	}
+
+	pub fn as_str(&self) -> &str
+	{
+		match *self {
+			Ty::Int8        => "int8_t",
+			Ty::Int16       => "int16_t",
+			Ty::Int32       => "int32_t",
+			Ty::Int64       => "int64_t",
+			Ty::Float       => "float",
+			Ty::Double      => "double",
+			Ty::String      => "string",
+			Ty::Boolean     => "boolean",
+			Ty::User(ref s) => s
+		}
+	}
+
+	fn size(&self) -> usize
+	{
+		match *self {
+			Ty::Int8    => ::std::mem::size_of::<i8>(),
+			Ty::Int16   => ::std::mem::size_of::<i16>(),
+			Ty::Int32   => ::std::mem::size_of::<i32>(),
+			Ty::Int64   => ::std::mem::size_of::<i64>(),
+			Ty::Float   => ::std::mem::size_of::<f32>(),
+			Ty::Double  => ::std::mem::size_of::<f64>(),
+			Ty::Boolean => ::std::mem::size_of::<i8>(),
+			_           => panic!("Tried to get fixed size of non-primitive"),
+		}
+	}
+
+	fn get_base_type(t: &syn::Type) -> Self
+	{
+		// There are two base types allowed here. The `Path` type contains all
+		// of the primitives and `Vec`. The `Array` type is fixed-size arrays.
+		match *t {
+			syn::Type::Path(syn::TypePath { path: syn::Path { ref segments, .. }, .. }) => {
+				// This is either a `Vec` or a primitive
+				match segments.iter().last().unwrap().ident.as_ref() {
+					"i8"     => Ty::Int8,
+					"i16"    => Ty::Int16,
+					"i32"    => Ty::Int32,
+					"i64"    => Ty::Int64,
+					"f32"    => Ty::Float,
+					"f64"    => Ty::Double,
+					"bool"   => Ty::Boolean,
+					"String" => Ty::String,
+					"Vec"    => Ty::get_base_type(get_vec_inner_type(t)),
+					n @ _    => Ty::User(n.to_string()),
+				}
+			},
+			syn::Type::Array(syn::TypeArray { ref elem, ..}) => {
+				// This is an array. Go a level deeper to figure out the base type
+				Ty::get_base_type(elem.as_ref())
+			},
+			_ => panic!("Type must either be an LCM primitive or an array"),
+		}
+	}
+}
+
+#[derive(Debug)]
+pub enum Dim
+{
+	Fixed(usize),
+	Variable(String),
+}
+impl Dim
+{
+	pub fn dim_type(&self) -> i8
+	{
+		match *self {
+			Dim::Fixed(_)    => 0,
+			Dim::Variable(_) => 1,
+		}
+	}
+
+	pub fn as_string(&self) -> String
```

Comment:
`Dim::as_string` should probably return a `Cow<'a, String>` to avoid the unnecessary clone.